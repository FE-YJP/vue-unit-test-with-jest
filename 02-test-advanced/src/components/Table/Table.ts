import Vue from 'vue'

export default Vue.extend({
  data: () => ({
    headers: [
      {
        text: 'Dessert (100g serving)',
        align: 'start',
        sortable: false,
        value: 'name'
      },
      { text: 'Calories', value: 'calories' },
      { text: 'Fat (g)', value: 'fat' },
      { text: 'Carbs (g)', value: 'carbs' },
      { text: 'Protein (g)', value: 'protein' },
      { text: 'Iron (%)', value: 'iron' }
    ],
    desserts: [
      {
        name: 'Frozen Yogurt',
        calories: 159,
        fat: 6.0,
        carbs: 24,
        protein: 4.0,
        iron: '1%'
      },
      {
        name: 'Ice cream sandwich',
        calories: 237,
        fat: 9.0,
        carbs: 37,
        protein: 4.3,
        iron: '1%'
      },
      {
        name: 'Eclair',
        calories: 262,
        fat: 16.0,
        carbs: 23,
        protein: 6.0,
        iron: '7%'
      },
      {
        name: 'Cupcake',
        calories: 305,
        fat: 3.7,
        carbs: 67,
        protein: 4.3,
        iron: '8%'
      },
      {
        name: 'Gingerbread',
        calories: 356,
        fat: 16.0,
        carbs: 49,
        protein: 3.9,
        iron: '16%'
      },
      {
        name: 'Jelly bean',
        calories: 375,
        fat: 0.0,
        carbs: 94,
        protein: 0.0,
        iron: '0%'
      },
      {
        name: 'Lollipop',
        calories: 392,
        fat: 0.2,
        carbs: 98,
        protein: 0,
        iron: '2%'
      },
      {
        name: 'Honeycomb',
        calories: 408,
        fat: 3.2,
        carbs: 87,
        protein: 6.5,
        iron: '45%'
      },
      {
        name: 'Donut',
        calories: 452,
        fat: 25.0,
        carbs: 51,
        protein: 4.9,
        iron: '22%'
      },
      {
        name: 'KitKat',
        calories: 518,
        fat: 26.0,
        carbs: 65,
        protein: 7,
        iron: '6%'
      }
    ],
    finallyData: '',
    error: ''
  }),

  methods: {
    async fetchData () {
      // 强烈推荐使用try catch 的写法
      try {
        // 下面也强烈推荐使用await的写法，代码可读性高，且易于测试
        const res = await this.$axios.get('/mockData.json')
        this.desserts = res.data?.mockData
      } catch (e) {
        // 此处一般需要根据错误码来执行不同的操作,或不操作
        this.error = 'FETCH_ERROR'
      } finally {
        // 此处不管请求成功与否都执行某操作（如取消loading）
        this.finallyData = 'finished'
      }
    }
  }
})
