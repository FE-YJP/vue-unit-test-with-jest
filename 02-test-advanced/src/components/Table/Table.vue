<template>
<div>
    <v-data-table
    :headers="headers"
    :items="desserts"
    :items-per-page="5"
    class="elevation-1"
  ></v-data-table>

   <v-btn
   @click="fetchData"
   class="mt-4"
   large
   depressed
   color="error"
   >
   点我获取mockData(axios测试用例)
   </v-btn>
</div>

</template>

<script lang="ts" src="./Table.ts"></script>
