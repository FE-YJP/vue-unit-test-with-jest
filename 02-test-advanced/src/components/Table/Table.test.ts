/**
 * axios测试的详解---使用nock进行测试（推荐）
 */

import { shallowMount } from '@vue/test-utils'
import Table from './Table.vue'
import nock from 'nock'; // https://github.com/nock/nock#how-does-it-work

describe('components.Table', () => {
  let wrapper: any

  // axios测试
  describe('methods.fetchData', () => {
    // 项目指定的baseUrl，一般测试时都是http://localhost
    const baseUrl ='http://localhost'

    afterEach(() => {
      nock.cleanAll();
    });

    it('正常获取数据', async () => {
      nock(baseUrl)
      .get('/mockData.json')
      .reply(200, {mockData:[]})
      .persist();

      wrapper = shallowMount(Table);
      await wrapper.vm.fetchData()
      // 验证正常获取nock的模拟数据
      expect(wrapper.vm.desserts).toEqual([])
      // 验证finally内操作执行
      expect(wrapper.vm.finallyData).toBe('finished')
    })


    it('请求出错时error赋值为FETCH_ERROR', async () => {
      nock(baseUrl)
      .get('/mockData.json')
      .reply(404, {})
      .persist();

      wrapper = shallowMount(Table)
      await wrapper.vm.fetchData()
      // 验证error的值
      expect(wrapper.vm.error).toBe('FETCH_ERROR')
      // 验证finally内操作执行
      expect(wrapper.vm.finallyData).toBe('finished')
    })
  })

  // 只要存在.vue文件，就必须进行snapshot测试
  describe('snapshot', () => {
    it('snapshot测试', () => {
      wrapper = shallowMount(Table)
      expect(wrapper.vm.$el).toMatchSnapshot();
    })
  })
})
