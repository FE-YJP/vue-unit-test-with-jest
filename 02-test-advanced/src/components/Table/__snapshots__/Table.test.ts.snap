// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components.Table snapshot snapshot测试 1`] = `
<div>
  <v-data-table-stub
    class="elevation-1"
    customfilter="function defaultFilter(value, search, item) {
  return value != null && search != null && typeof value !== 'boolean' && value.toString().toLocaleLowerCase().indexOf(search.toLocaleLowerCase()) !== -1;
}"
    customgroup="function groupItems(items, groupBy, groupDesc) {
  var key = groupBy[0];
  var groups = [];
  var current = null;

  for (var i = 0; i < items.length; i++) {
    var item = items[i];
    var val = getObjectValueByPath(item, key);

    if (current !== val) {
      current = val;
      groups.push({
        name: val,
        items: []
      });
    }

    groups[groups.length - 1].items.push(item);
  }

  return groups;
}"
    customsort="function sortItems(items, sortBy, sortDesc, locale, customSorters) {
  if (sortBy === null || !sortBy.length) return items;
  var stringCollator = new Intl.Collator(locale, {
    sensitivity: 'accent',
    usage: 'sort'
  });
  return items.sort(function (a, b) {
    var _a, _b;

    for (var i = 0; i < sortBy.length; i++) {
      var sortKey = sortBy[i];
      var sortA = getObjectValueByPath(a, sortKey);
      var sortB = getObjectValueByPath(b, sortKey);

      if (sortDesc[i]) {
        _a = __read([sortB, sortA], 2), sortA = _a[0], sortB = _a[1];
      }

      if (customSorters && customSorters[sortKey]) {
        var customResult = customSorters[sortKey](sortA, sortB);
        if (!customResult) continue;
        return customResult;
      } // Check if both cannot be evaluated


      if (sortA === null && sortB === null) {
        continue;
      }

      _b = __read([sortA, sortB].map(function (s) {
        return (s || '').toString().toLocaleLowerCase();
      }), 2), sortA = _b[0], sortB = _b[1];

      if (sortA !== sortB) {
        if (!isNaN(sortA) && !isNaN(sortB)) return Number(sortA) - Number(sortB);
        return stringCollator.compare(sortA, sortB);
      }
    }

    return 0;
  });
}"
    expanded=""
    expandicon="$expand"
    groupby=""
    groupdesc=""
    headers="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
    itemkey="id"
    items="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
    itemsperpage="5"
    loadingtext="$vuetify.dataIterator.loadingText"
    locale="en-US"
    mobilebreakpoint="600"
    nodatatext="$vuetify.noDataText"
    noresultstext="$vuetify.dataIterator.noResultsText"
    options="[object Object]"
    page="1"
    selectablekey="isSelectable"
    serveritemslength="-1"
    sortby=""
    sortdesc=""
    value=""
  />
   
  <v-btn-stub
    activeclass=""
    class="mt-4"
    color="error"
    depressed="true"
    large="true"
    tag="button"
    type="button"
  >
    
   点我获取mockData(axios测试用例)
   
  </v-btn-stub>
</div>
`;
